src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -1,16 +1,1108 @@
+                            builder.AppendLine("await System.Threading.Tasks.Task.CompletedTask;");
+                            break;
+                        case YieldBreak:
+                            //yield break;
+                            builder.AppendLine($@"yield break;");
+                            break;
+                    }
+                }
+                expectationBuilder.Append("Done");
+                return (builder.ToString(), expectationBuilder.ToString());
+            }
+        }
+
+        [Fact]
+        public void AsyncIteratorWithAwaitAndYieldAndAwait()
+        {
+            string source = @"
+using static System.Console;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        Write(""1 "");
+        await System.Threading.Tasks.Task.Delay(10);
+        Write(""2 "");
+        yield return 3;
+        Write(""4 "");
+        await System.Threading.Tasks.Task.Delay(10);
+    }
+    static async System.Threading.Tasks.Task Main()
+    {
+        Write(""0 "");
+        foreach await (var i in M())
+        {
+            Write($""{i} "");
+        }
+        Write(""Done"");
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "0 1 2 3 4 Done");
+        }
+
+        [Fact]
+        public void AsyncIteratorWithAwaitOnly()
+        {
+            string source = @"
+class C
+{
+    async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        await System.Threading.Tasks.Task.CompletedTask;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common });
+            comp.VerifyDiagnostics(
+                // (4,60): error CS0161: 'C.M()': not all code paths return a value
+                //     async System.Collections.Generic.IAsyncEnumerable<int> M()
+                Diagnostic(ErrorCode.ERR_ReturnExpected, "M").WithArguments("C.M()").WithLocation(4, 60)
+                );
+        }
+
+        [Fact]
+        public void AsyncIteratorWithYieldOnly()
+        {
+            string source = @"
+class C
+{
+    async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        yield return 1;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common });
+            comp.VerifyDiagnostics(
+                // (4,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerable<int> M()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(4, 60)
+                );
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider looping over async enumerable to verify state machine returns one value.

Reply:
Please add a test for a single `yield break;` with on preceding `await`.

---
In reply to: [215014459](https://github.com/dotnet/roslyn/pull/28218#discussion_r215014459) [](ancestors = 215014459)

Reply:
The second scenario (just `yield break;` is covered in `AsyncIteratorWithCustomCode`):

```
  verify(new[] { AwaitFast, YieldBreak });
  verify(new[] { AwaitSlow, YieldBreak });
```

---
In reply to: [215016493](https://github.com/dotnet/roslyn/pull/28218#discussion_r215016493) [](ancestors = 215016493,215014459)